### ISSUE TITLE
Cleanup the remote private_data_dir on execution nodes, from jobs ran on those nodes

### ISSUE THREAD

**Issue by AlanCoding** [ansible/awx#10687]

**Comment by jladdjr** [comment]
I don't think we actually have anything to clean up on the remote nodes. When the execution node goes to run the job, it spins up an execution environment. Whenever runner does this, [it appends `--rm`](https://github.com/ansible/ansible-runner/blob/fe3c1aa865898d4a173079ea846f9af5adc79e32/ansible_runner/config/_base.py#L425) to the `<container runtime> run` command. This means that as soon as the container finishes running it will get cleaned up.

Motion to close @AlanCoding

**Comment by jladdjr** [comment]
The plot thickens. While the container is cleaned up, the private_data_dir is _mounted_ to the container and sticks around even after the container disappears.

A job details page will show this in action. `job_args` shows something like:
```
"job_args": "[\"podman\", \"run\", \"--rm\", \"--tty\", \"--interactive\", \"--workdir\", \"/runner/project\", \"-v\", \"/tmp/pdd_wrapper_6_5vfv3t74/awx_6_pgh5064a/:/runner/:Z\", \"--env-file\", \"/tmp/pdd_wrapper_6_5vfv3t74/awx_6_pgh5064a/artifacts/6/env.list\", \"--quiet\", \"--name\", \"ansible_runner_6\", \"--user=root\", \"quay.io/ansible/awx-ee:latest\", \"ssh-agent\", \"sh\", \"-c\", \"trap 'rm -f /runner/artifacts/6/ssh_key_data' EXIT && ssh-add /runner/artifacts/6/ssh_key_data && rm -f /runner/artifacts/6/ssh_key_data && ansible-playbook -u root --ask-pass --ask-become-pass -i /runner/inventory/hosts -e @/runner/env/extravars ping.yml\"]",
```

.. which includes an option to map the private data dir into the container:
`... "-v\", \"/tmp/pdd_wrapper_6_5vfv3t74/awx_6_pgh5064a/:/runner/:Z\" ...`

Ran a job agains an execution node and found that `/tmp/pdd_wrapper_6_5vfv3t74` was left after a job was run. `/tmp/pdd_wrapper_6_5vfv3t74/awx_6_pgh5064a/env/` had sensitive data (in files like passwords and ssh_key) that we would def. want to clean up.

**Comment by AlanCoding** [comment]
People have definitely written code in runner to clean up the private data dir

https://github.com/ansible/ansible-runner/blob/77f15a45b150dcd5b43d6a3fb0aec77c31f7f688/ansible_runner/__main__.py#L792-L794

this is for worker.

However, that's for `if not vargs.get('private_data_dir'):` (it creates its own temporary dir), and AWX definitely passes the private_data_dir.

We need changes in runner so that we can tell worker to delete the private_data_dir when it finishes. Honestly, it's kind of weird that it doesn't, because `ansible-runner worker` _writes_ it after unzipping the streamed data in the first place.

**Comment by jladdjr** [comment]
👍  I was just tinkering with https://github.com/ansible/ansible-runner/compare/devel...jladdjr:ansible_process_cleanup_private_data_dir?expand=1 to make sure that awx grabbed everything it needed from the job (via callbacks) and we could safely cleanup the dir immediately after ansible-runner process finished.

**Comment by AlanCoding** [comment]
We have to be cautious that errors thrown _after_ unpacking the directory but before the job finishes won't prevent this from being ran. The directory is written by `unstream_dir`, but you could even have an error _halfway_ through `unstream_dir` after secrets are written.

For these problems, `@atexit.register` really is convenient. If you look at `Worker.__init__`, where it sets `private_data_dir = kwargs.get('private_data_dir')`, right after _that_ is where I'd want to register a method to cleanup the private_data_dir, because it's the earliest moment where it's possible.

**Comment by jladdjr** [comment]
I tried that and must be missing something because the directory doesn't get cleaned up. I've run a bunch of sanity checks and it seems like:

* atexit is working
* the shutil.rmtree call is run
* .. but doesn't seem to complete (subsequent code not run)
* even so, there's no exception caught, and `rmtree` never seems to even call the error handler I provided

It's as if the runner process itself is being reaped. I'm going to pause here, but would love another set of 👀  on this to help restore sanity. It certainly seems like this should be cleaning up the directory.

---

oh, and fwiw, when I call `rmtree` in a standalone python session, the directory gets cleaned up w/out issues.
